File: backend/src/main/java/t/lab/guide/dto/admin/point/AdminPatchPointRequest.java
package t.lab.guide.dto.admin.point;

import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.AssertTrue;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.Size;
import t.lab.guide.dto.common.GeoPoint;

public record AdminPatchPointRequest(
        @Schema(description = "Новое название точки интереса",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                maxLength = 255,
                example = "Красная площадь")
        @Size(max = 255)
        String title,

        @Schema(description = "Новое подробное описание точки интереса",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                maxLength = 5000,
                example = "Красная площадь — главная площадь Москвы.")
        @Size(max = 5000)
        String description,

        @Schema(description = "Новый идентификатор категории",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                example = "2")
        Long categoryId,

        @Schema(description = "Новый адрес точки интереса",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                maxLength = 500,
                example = "Москва, Красная площадь, 1")
        @Size(max = 500)
        String address,

        @Schema(description = "Новые координаты точки",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED)
        @Valid
        GeoPoint coordinates,

        @Schema(description = "Новое среднее время посещения точки в минутах",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                example = "90")
        @Min(1)
        Integer visitTime,

        @Schema(description = "Новые часы работы точки",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                maxLength = 255,
                example = "10:00 - 22:00")
        @Size(max = 255)
        String workingHours,

        @Schema(description = "Активна ли точка (показывается пользователям)",
                requiredMode = Schema.RequiredMode.NOT_REQUIRED,
                example = "true")
        Boolean isActive
) {
    @AssertTrue(message = "Необходимо передать хотя бы одно поле для обновления")
    @Schema(hidden = true)
    public boolean isAnyFieldPresent() {
        return title != null
                || description != null
                || categoryId != null
                || address != null
                || coordinates != null
                || visitTime != null
                || workingHours != null
                || isActive != null;
    }
}

File: backend/src/main/kotlin/t/lab/guide/dto/admin/point/AdminPatchPointRequest.kt
package t.lab.guide.dto.admin.point

import com.fasterxml.jackson.annotation.JsonIgnore
import io.swagger.v3.oas.annotations.media.Schema
import jakarta.validation.Valid
import jakarta.validation.constraints.AssertTrue
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.Size
import t.lab.guide.dto.common.GeoPoint

@Schema(
    description =
        "Запрос на частичное обновление точки интереса администратором. " +
            "Передаются только изменяемые поля; минимум одно поле обязательно.",
)
data class AdminPatchPointRequest(
    @Schema(
        description = "Новое название точки интереса",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        maxLength = 255,
        example = "Красная площадь",
    )
    @field:Size(max = 255, message = "максимальная длина названия - 255 символов!")
    val title: String? = null,
    @Schema(
        description = "Новое подробное описание точки интереса",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        maxLength = 5000,
        example = "Красная площадь — главная площадь Москвы.",
    )
    @field:Size(max = 5000, message = "максимальная длина описания - 5000 символов!")
    val description: String? = null,
    @Schema(
        description = "Новое краткое описание точки интереса",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        maxLength = 255,
        example = "Главная площадь Москвы",
    )
    @field:Size(max = 255, message = "максимальная длина краткого описания - 255 символов!")
    val shortDescription: String? = null,
    @Schema(
        description = "Новый идентификатор категории",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        example = "2",
    )
    val categoryId: Long? = null,
    @Schema(
        description = "Новый адрес точки интереса",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        maxLength = 255,
        example = "Москва, Красная площадь, 1",
    )
    @field:Size(max = 255, message = "максимальная длина адреса - 255 символов!")
    val address: String? = null,
    @Schema(
        description = "Новые координаты точки",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
    )
    @field:Valid
    val coordinates: GeoPoint? = null,
    @Schema(
        description = "Новое среднее время посещения точки в минутах",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        example = "90",
    )
    @field:Min(1, message = "время посещения должно быть положительным числом!")
    val visitTime: Int? = null,
    @Schema(
        description = "Новые часы работы точки",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        maxLength = 255,
        example = "10:00 - 22:00",
    )
    @field:Size(max = 255, message = "максимальная длина часов работы - 255 символов!")
    val workingHours: String? = null,
    @Schema(
        description = "Активна ли точка (показывается пользователям)",
        requiredMode = Schema.RequiredMode.NOT_REQUIRED,
        example = "true",
    )
    val isActive: Boolean? = null,
) {
    @AssertTrue(message = "Необходимо передать хотя бы одно поле для обновления")
    @JsonIgnore
    @Schema(hidden = true)
    fun isAnyFieldPresent(): Boolean =
        title != null ||
            description != null ||
            categoryId != null ||
            address != null ||
            coordinates != null ||
            visitTime != null ||
            workingHours != null ||
            isActive != null
}
